snapshotter/devmapper/metadata.go

```diff
@@ -0,0 +1,314 @@
+	maxDeviceID = 0xffffff // Device IDs are 24-bit numbers
+	deviceFree  = byte(0)
+	deviceTaken = byte(1)
+)
+
+// Bucket names
+var (
+	devicesBucketName  = []byte("devices")    // Contains thin devices metadata <device_name>=<DeviceInfo>
+	deviceIDBucketName = []byte("device_ids") // Tracks used device ids <device_id_[0..maxDeviceID)>=<byte_[0/1]>
+)
+
+var (
+	ErrNotFound      = errors.New("not found")
+	ErrAlreadyExists = errors.New("object already exists")
+)
+
+// PoolMetadata keeps device info for the given thin-pool device, it also responsible for
+// generating next available device ids and tracking devmapper transaction numbers
+type PoolMetadata struct {
+	db *bolt.DB
+}
+
+func NewPoolMetadata(dbfile string) (*PoolMetadata, error) {
+	db, err := bolt.Open(dbfile, 0600, nil)
+	if err != nil {
+		return nil, err
+	}
+
+	metadata := &PoolMetadata{db: db}
+	if err := metadata.ensureDatabaseInitialized(); err != nil {
+		return nil, errors.Wrap(err, "failed to initialize database")
+	}
+
+	return metadata, nil
+}
+
+func (m *PoolMetadata) ensureDatabaseInitialized() error {
+	return m.db.Update(func(tx *bolt.Tx) error {
+		if _, err := tx.CreateBucketIfNotExists(devicesBucketName); err != nil {
+			return err
+		}
+
+		if _, err := tx.CreateBucketIfNotExists(deviceIDBucketName); err != nil {
+			return err
+		}
+
+		return nil
+	})
+}
+
+// AddDevice saves device info to database.
+// Callback should be used to acquire device ID or to rollback transaction in case of error.
+func (m *PoolMetadata) AddDevice(ctx context.Context, info *DeviceInfo, fn DeviceIDCallback) error {
+	return m.db.Update(func(tx *bolt.Tx) error {
+		devicesBucket := tx.Bucket(devicesBucketName)
+
+		// Make sure device name is unique
+		if err := getObject(devicesBucket, info.Name, nil); err == nil {
+			return ErrAlreadyExists
+		}
+
+		// Find next available device ID
+		deviceID, err := getNextDeviceID(tx)
+		if err != nil {
+			return err
+		}
+
+		if err := fn(deviceID); err != nil {
+			return err
+		}
+
+		info.DeviceID = deviceID
+
+		if err := putObject(devicesBucket, info.Name, info, false); err != nil {
+			return err
+		}
+
+		return nil
+	})
+}
+
+func getNextDeviceID(tx *bolt.Tx) (int, error) {
+	bucket := tx.Bucket(deviceIDBucketName)
+	cursor := bucket.Cursor()
+
+	// Check if any device id can be reused.
+	// Bolt stores its keys in byte-sorted order within a bucket.
+	// This makes sequential iteration extremely fast.
+	for key, taken := cursor.First(); key != nil; key, taken = cursor.Next() {
+		isFree := taken[0] == deviceFree
+		if !isFree {
+			continue
+		}
+
+		id, err := strconv.Atoi(string(key))
+		if err != nil {
+			return 0, err
+		}
+
+		if err := markDeviceID(tx, id, deviceTaken); err != nil {
+			return 0, err
+		}
+
+		return id, nil
+	}
+
+	// Try allocate new device ID
+	seq, err := bucket.NextSequence()
+	if err != nil {
+		return 0, err
+	}
+
+	if seq >= maxDeviceID {
+		return 0, errors.Errorf("couldn't find free device key")
```

Comment:
Can you prefix this error with a short string to identify the package so it's easier to find the source code when something goes wrong?  Something like "dm-meta: " or "dm-snapshotter: "?